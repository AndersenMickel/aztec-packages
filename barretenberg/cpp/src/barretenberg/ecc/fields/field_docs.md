Prime field documentation    {#field_docs}
===
Barretenberg has its own implementation of finite field arithmetic. The implementation targets 254 (bn254, grumpkin) and 256-bit (secp256k1, secp256r1) fields. Internally the field is representate as a little-endian C-array of 4 uint64_t limbs.

## Field arithmetic
We use Montgomery reduction to speed up field multiplication. For an original element  \f$ a ∈ F_p\f$ the element is represented internally as \f$ a⋅R\ mod\ p\f$, where \f$R = 2^k\ mod\ p\f$. The chosen \f$k\f$ depends on the build configurations. For builds that don't support the uint128_t type, for example, for WASM build, \f$k=29⋅9=261\f$, while for standard builds the value of \f$k=64⋅4=256\f$. The idea behind this form is to avoid heavy division modulo \f$p\f$. To compute a representative of element \f$c = a⋅b\ mod\ p\f$ we compute \f$c⋅R = (a⋅R)⋅(b⋅R) / R\ mod\ p\f$, but we use an efficient division trick to avoid straight modular division (explanation for the 4⋅64 case):
1. First, we compute the value \f$c_r=c⋅R\f$ in integers and get a value with 8 64-bit limbs
2. Then we take the lowest limb of \f$c_r\f$ (\f$c_r[0]\f$) and multiply it by a special value \f$r_{inv} = -1 ⋅ p^{-1}\ mod\  2^{64}\f$, getting \f$k = r_{inv}⋅ c_r[0]\ mod\ 2^{64}\f$
3. Next we update \f$c_r\f$ in integers by adding a value \f$k⋅p\f$: \f$c_r += k⋅p\f$. You might notice that the value of \f$c_r\ mod\ p\f$ hasn't changes, since we've added a multiple of the modulus. A the same time, if we look at the expression modulo \f$2^{64}\f$: \f$c_r + k⋅p = c_r + c_r⋅r_{inv}⋅p = c_r + c_r⋅ (-1)⋅p^{-1}⋅p = c_r - c_r = 0\ mod\ 2^{64}\f$
4. We perform the same operation for \f$c_r[1]\f$, but instead of adding \f$k⋅p\f$, we add \f$2^{64}⋅k⋅p\f$. In the implementation, instead of adding $k⋅ p$ to limbs of $c_r$ starting with zero, we just start with limb number 1. This ensures that \f$c_r[1]=0\f$. We then perform the same operation for 2 more limbs.


This is my field documentation \f$\frac{1}{2}= 0\f$ alalalalalalala

## Helpful python snippets

Parse field parameters
```python
print(help)
```