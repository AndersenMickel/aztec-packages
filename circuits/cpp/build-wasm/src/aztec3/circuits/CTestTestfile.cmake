# CMake generated Testfile for 
# Source directory: /mnt/user-data/mara/aztec-packages/circuits/cpp/src/aztec3/circuits
# Build directory: /mnt/user-data/mara/aztec-packages/circuits/cpp/build-wasm/src/aztec3/circuits
# 
# This file includes the relevant testing commands required for 
# testing this directory and lists subdirectories to be tested as well.
subdirs("apps")
subdirs("abis")
subdirs("kernel")
subdirs("rollup")
subdirs("recursion")
