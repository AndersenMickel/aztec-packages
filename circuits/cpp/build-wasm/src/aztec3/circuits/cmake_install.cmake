# Install script for directory: /mnt/user-data/mara/aztec-packages/circuits/cpp/src/aztec3/circuits

# Set the install prefix
if(NOT DEFINED CMAKE_INSTALL_PREFIX)
  set(CMAKE_INSTALL_PREFIX "/mnt/user-data/mara/aztec-packages/circuits/cpp/barretenberg/cpp/src/wasi-sdk-20.0/share/wasi-sysroot")
endif()
string(REGEX REPLACE "/$" "" CMAKE_INSTALL_PREFIX "${CMAKE_INSTALL_PREFIX}")

# Set the install configuration name.
if(NOT DEFINED CMAKE_INSTALL_CONFIG_NAME)
  if(BUILD_TYPE)
    string(REGEX REPLACE "^[^A-Za-z0-9_]+" ""
           CMAKE_INSTALL_CONFIG_NAME "${BUILD_TYPE}")
  else()
    set(CMAKE_INSTALL_CONFIG_NAME "Release")
  endif()
  message(STATUS "Install configuration: \"${CMAKE_INSTALL_CONFIG_NAME}\"")
endif()

# Set the component getting installed.
if(NOT CMAKE_INSTALL_COMPONENT)
  if(COMPONENT)
    message(STATUS "Install component: \"${COMPONENT}\"")
    set(CMAKE_INSTALL_COMPONENT "${COMPONENT}")
  else()
    set(CMAKE_INSTALL_COMPONENT)
  endif()
endif()

# Is this installation the result of a crosscompile?
if(NOT DEFINED CMAKE_CROSSCOMPILING)
  set(CMAKE_CROSSCOMPILING "TRUE")
endif()

# Set default install directory permissions.
if(NOT DEFINED CMAKE_OBJDUMP)
  set(CMAKE_OBJDUMP "/mnt/user-data/mara/aztec-packages/circuits/cpp/barretenberg/cpp/src/wasi-sdk-20.0/bin/llvm-objdump")
endif()

if(NOT CMAKE_INSTALL_LOCAL_ONLY)
  # Include the install script for the subdirectory.
  include("/mnt/user-data/mara/aztec-packages/circuits/cpp/build-wasm/src/aztec3/circuits/apps/cmake_install.cmake")
endif()

if(NOT CMAKE_INSTALL_LOCAL_ONLY)
  # Include the install script for the subdirectory.
  include("/mnt/user-data/mara/aztec-packages/circuits/cpp/build-wasm/src/aztec3/circuits/abis/cmake_install.cmake")
endif()

if(NOT CMAKE_INSTALL_LOCAL_ONLY)
  # Include the install script for the subdirectory.
  include("/mnt/user-data/mara/aztec-packages/circuits/cpp/build-wasm/src/aztec3/circuits/kernel/cmake_install.cmake")
endif()

if(NOT CMAKE_INSTALL_LOCAL_ONLY)
  # Include the install script for the subdirectory.
  include("/mnt/user-data/mara/aztec-packages/circuits/cpp/build-wasm/src/aztec3/circuits/rollup/cmake_install.cmake")
endif()

if(NOT CMAKE_INSTALL_LOCAL_ONLY)
  # Include the install script for the subdirectory.
  include("/mnt/user-data/mara/aztec-packages/circuits/cpp/build-wasm/src/aztec3/circuits/recursion/cmake_install.cmake")
endif()

