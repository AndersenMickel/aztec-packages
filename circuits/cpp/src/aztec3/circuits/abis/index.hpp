#include "call_context.hpp"
#include "call_stack_item.hpp"
#include "function_data.hpp"
#include "private_circuit_public_inputs.hpp"
#include "storage_read.hpp"
#include "storage_write.hpp"