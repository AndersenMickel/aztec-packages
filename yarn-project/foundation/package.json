{
  "name": "@aztec/foundation",
  "version": "0.1.0",
  "packageManager": "yarn@3.4.1",
  "type": "module",
  "main": "./dest/index.js",
  "types": "./dest/index.d.ts",
  "exports": {
    "./eslint": "./.eslintrc.cjs",
    "./eslint.docs": "./.eslintrc.docs.cjs",
    "./prettier": "./.prettierrc.json",
    "./abi": {
      "bun": "./src/abi/index.ts",
      "default": "./dest/abi/index.js"
    },
    "./async-map": {
      "bun": "./src/async-map/index.ts",
      "default": "./dest/async-map/index.js"
    },
    "./aztec-address": {
      "bun": "./src/aztec-address/index.ts",
      "default": "./dest/aztec-address/index.js"
    },
    "./collection": {
      "bun": "./src/collection/index.ts",
      "default": "./dest/collection/index.js"
    },
    "./crypto": {
      "bun": "./src/crypto/index.ts",
      "default": "./dest/crypto/index.js"
    },
    "./error": {
      "bun": "./src/error/index.ts",
      "default": "./dest/error/index.js"
    },
    "./eth-address": {
      "bun": "./src/eth-address/index.ts",
      "default": "./dest/eth-address/index.js"
    },
    "./fifo": {
      "bun": "./src/fifo/index.ts",
      "default": "./dest/fifo/index.js"
    },
    "./json-rpc": {
      "bun": "./src/json-rpc/index.ts",
      "default": "./dest/json-rpc/index.js"
    },
    "./json-rpc/server": {
      "bun": "./src/json-rpc/server/index.ts",
      "default": "./dest/json-rpc/server/index.js"
    },
    "./json-rpc/client": {
      "bun": "./src/json-rpc/client/index.ts",
      "default": "./dest/json-rpc/client/index.js"
    },
    "./log": {
      "bun": "./src/log/index.ts",
      "default": "./dest/log/index.js"
    },
    "./mutex": {
      "bun": "./src/mutex/index.ts",
      "default": "./dest/mutex/index.js"
    },
    "./fields": {
      "bun": "./src/fields/index.ts",
      "default": "./dest/fields/index.js"
    },
    "./retry": {
      "bun": "./src/retry/index.ts",
      "default": "./dest/retry/index.js"
    },
    "./running-promise": {
      "bun": "./src/running-promise/index.ts",
      "default": "./dest/running-promise/index.js"
    },
    "./serialize": {
      "bun": "./src/serialize/index.ts",
      "default": "./dest/serialize/index.js"
    },
    "./sleep": {
      "bun": "./src/sleep/index.ts",
      "default": "./dest/sleep/index.js"
    },
    "./timer": {
      "bun": "./src/timer/index.ts",
      "default": "./dest/timer/index.js"
    },
    "./transport": {
      "bun": "./src/transport/index.ts",
      "default": "./dest/transport/index.js"
    },
    "./trees": {
      "bun": "./src/trees/index.ts",
      "default": "./dest/trees/index.js"
    },
    "./wasm": {
      "bun": "./src/wasm/index.ts",
      "default": "./dest/wasm/index.js"
    },
    "./worker": {
      "bun": "./src/worker/index.ts",
      "default": "./dest/worker/index.js"
    },
    "./bigint-buffer": {
      "bun": "./src/bigint-buffer/index.ts",
      "default": "./dest/bigint-buffer/index.js"
    },
    "./types": {
      "bun": "./src/types/index.ts",
      "default": "./dest/types/index.js"
    },
    "./url": {
      "bun": "./src/url/index.ts",
      "default": "./dest/url/index.js"
    },
    "./committable": {
      "bun": "./src/committable/index.ts",
      "default": "./dest/committable/index.js"
    },
    "./noir": {
      "bun": "./src/noir/index.ts",
      "default": "./dest/noir/index.js"
    },
    "./testing": {
      "bun": "./src/testing/index.ts",
      "default": "./dest/testing/index.js"
    },
    "./array": {
      "bun": "./src/array/index.ts",
      "default": "./dest/array/index.js"
    },
    "./validation": {
      "bun": "./src/validation/index.ts",
      "default": "./dest/validation/index.js"
    }
  },
  "scripts": {
    "build": "yarn clean && tsc -b",
    "build:dev": "tsc -b --watch",
    "clean": "rm -rf ./dest .tsbuildinfo",
    "formatting": "run -T prettier --check ./src && run -T eslint ./src",
    "formatting:fix": "run -T eslint --fix ./src && run -T prettier -w ./src",
    "test:light": "NODE_NO_WARNINGS=1 node --experimental-vm-modules $(yarn bin jest) --passWithNoTests",
    "test": "NODE_NO_WARNINGS=1 node --experimental-vm-modules $(yarn bin jest) --passWithNoTests"
  },
  "inherits": [
    "../package.common.json"
  ],
  "jest": {
    "preset": "ts-jest/presets/default-esm",
    "moduleNameMapper": {
      "^(\\.{1,2}/.*)\\.[cm]?js$": "$1"
    },
    "testRegex": "./src/.*\\.test\\.(js|mjs|ts)$",
    "rootDir": "./src"
  },
  "dependencies": {
    "@aztec/bb.js": "portal:../../barretenberg/ts",
    "@koa/cors": "^4.0.0",
    "@noble/curves": "^1.2.0",
    "bn.js": "^5.2.1",
    "debug": "^4.3.4",
    "detect-node": "^2.1.0",
    "elliptic": "^6.5.4",
    "hash.js": "^1.1.7",
    "koa": "^2.14.2",
    "koa-bodyparser": "^4.4.0",
    "koa-compress": "^5.1.0",
    "koa-router": "^12.0.0",
    "leveldown": "^6.1.1",
    "levelup": "^5.1.1",
    "lodash.chunk": "^4.2.0",
    "lodash.clonedeepwith": "^4.5.0",
    "memdown": "^6.1.1",
    "pako": "^2.1.0",
    "sha3": "^2.1.4",
    "ts-essentials": "^9.4.1",
    "zod": "^3.22.4"
  },
  "devDependencies": {
    "@jest/globals": "^29.5.0",
    "@types/bn.js": "^5.1.3",
    "@types/debug": "^4.1.7",
    "@types/detect-node": "^2.0.0",
    "@types/elliptic": "^6.4.16",
    "@types/jest": "^29.5.0",
    "@types/koa": "^2.13.5",
    "@types/koa-bodyparser": "^4.3.10",
    "@types/koa-compress": "^4.0.3",
    "@types/koa-cors": "^0.0.2",
    "@types/koa-router": "^7.4.4",
    "@types/koa__cors": "^4.0.0",
    "@types/leveldown": "^4.0.3",
    "@types/levelup": "^5.1.2",
    "@types/lodash.chunk": "^4.2.9",
    "@types/lodash.clonedeepwith": "^4.5.7",
    "@types/memdown": "^3.0.1",
    "@types/node": "^18.7.23",
    "@types/pako": "^2.0.0",
    "@types/supertest": "^2.0.12",
    "@typescript-eslint/eslint-plugin": "^6.2.1",
    "@typescript-eslint/parser": "^6.2.1",
    "comlink": "^4.4.1",
    "eslint": "^8.21.0",
    "eslint-config-prettier": "^8.5.0",
    "eslint-plugin-jsdoc": "^40.1.0",
    "eslint-plugin-no-only-tests": "^3.1.0",
    "eslint-plugin-tsdoc": "^0.2.17",
    "jest": "^29.5.0",
    "prettier": "^2.7.1",
    "supertest": "^6.3.3",
    "ts-jest": "^29.1.0",
    "ts-node": "^10.9.1",
    "typescript": "^5.0.4"
  },
  "files": [
    "dest",
    "src",
    "!*.test.*"
  ],
  "engines": {
    "node": ">=18"
  }
}
