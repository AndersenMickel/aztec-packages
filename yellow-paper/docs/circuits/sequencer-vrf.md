# Sequencer VRF Circuit

The Sequencer VRF Circuit is intended to verify the VRF score submitted by the leading sequencer who built the block. It is executed during the proving phase and recursively verified by the root rollup circuit.

<!-- Q: Should this be a separate circuit, or should it be a constraint included in the root rollup circuit? -->

## Requirements

The sequencer VRF circuits constrains that the resulting score for a given sequencer with a given random seed, obtained from a RANDAO value, is equal to the reported score. The VRF is calculated as the Pedersen hash of the sequencer's identifier and the random seed.

## Private Inputs

None. All inputs to the circuit are public, as it is executed by a public prover.

## Public Inputs

- Sequencer address.
- Random seed.
- Reported VRF score.