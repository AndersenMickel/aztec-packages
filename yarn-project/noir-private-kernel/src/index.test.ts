import { makePrivateKernelInputsInit } from '@aztec/circuits.js';
import { DebugLogger, createDebugLogger } from '@aztec/foundation/log';
import { Fr } from '@aztec/foundation/fields';

import { executeInit } from './index.js';

describe('Private kernel', () => {
  let logger: DebugLogger;
  beforeAll(() => {
    logger = createDebugLogger('noir-private-kernel');
  });

  it('Executes private kernel init circuit with abi all zeroes (does not crash)', async () => {
    logger('Initialized Noir instance with private kernel init circuit');

    const kernelInputs = makePrivateKernelInputsInit();

    // TODO(Kev): How was this passing with the circuits.wasm version? Maybe it never tested init with mock?
    // Delegate and static calls are not allowed in the init
    kernelInputs.privateCall.callStackItem.publicInputs.callContext.isDelegateCall = false;
    kernelInputs.privateCall.callStackItem.publicInputs.callContext.isStaticCall = false;

    // Tx origin address must equal the storage contract address on the call stack
    kernelInputs.privateCall.callStackItem.contractAddress = kernelInputs.privateCall.callStackItem.publicInputs.callContext.storageContractAddress;
    kernelInputs.txRequest.origin = kernelInputs.privateCall.callStackItem.contractAddress;

    // Tx request function data and args hash must match
    kernelInputs.txRequest.functionData = kernelInputs.privateCall.callStackItem.functionData;
    kernelInputs.txRequest.argsHash = kernelInputs.privateCall.callStackItem.publicInputs.argsHash;

    // Make these 0 so that the circuit does not try to apply constraints on them
    kernelInputs.privateCall.callStackItem.publicInputs.privateCallStack.fill(Fr.ZERO);
    kernelInputs.privateCall.callStackItem.publicInputs.readRequests.fill(Fr.ZERO);
    
    const _kernelOutputs = await executeInit(kernelInputs);

    logger('Executed private kernel init circuit with all zeroes');
  });
});
