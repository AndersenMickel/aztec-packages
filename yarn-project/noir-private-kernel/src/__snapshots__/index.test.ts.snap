// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Noir compatibility tests (interop_testing.nr) Complete Address matches Noir 1`] = `
CompleteAddress {
  "address": AztecAddress {
    "buffer": {
      "data": [
        26,
        110,
        137,
        176,
        52,
        71,
        135,
        19,
        199,
        169,
        241,
        199,
        127,
        184,
        10,
        249,
        149,
        247,
        8,
        246,
        242,
        8,
        191,
        53,
        43,
        77,
        218,
        33,
        36,
        115,
        145,
        9,
      ],
      "type": "Buffer",
    },
  },
  "partialAddress": Fr {
    "value": 11517109422981783267562923429265051163864675433397402239476609863059550730459n,
  },
  "publicKey": Point {
    "kind": "point",
    "x": Fr {
      "value": 1n,
    },
    "y": Fr {
      "value": 2n,
    },
  },
}
`;

exports[`Noir compatibility tests (interop_testing.nr) ComputeContractAddressFromPartial matches Noir 1`] = `"0x0b487ff2900ae1178e131bfe333fdbc351beef658f7c0d62db2801429b1aab75"`;

exports[`Noir compatibility tests (interop_testing.nr) TxRequest Hash matches Noir 1`] = `"0x0b487ff2900ae1178e131bfe333fdbc351beef658f7c0d62db2801429b1aab75"`;
